### PR TITLE
fix(Core): Hunter Readiness crash fix

### PR DESCRIPTION
**Pull request by ZhengPeiRu21**
## Changes Proposed:
-  Prevent crash in some edge case of hunter use Readiness

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/12980
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/12971

## Tests Performed:

- None

## How to Test the Changes:

1. See if still crash in original situation

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Review by acidmanifesto — APPROVED** [review]

**Comment by Nefertumm** [comment]
itr will always be the same value now, causing a freeze.

**Comment by ZhengPeiRu21** [comment]
I have updated to a different way that should work better. Thanks Nefertum!

**Comment by fullofstarschen** [comment]
I tried [f6001fd], but failed.@ZhengPeiRu21

**Comment by Nefertumm** [comment]
@fullofstarschen try again, and please compile with relwithdebinfo

**Comment by MegaserverRU** [comment]
> @fullofstarschen try again.
I checked with a new script, everything works well in the Citadel of the Ice Crown, I also checked on mannequins and everything works well there. If I find something, I'll let you know.

https://github.com/azerothcore/azerothcore-wotlk/pull/12987/commits/6123d5c6ff230b88bb26da91119ce25fb21d3bb8

**Comment by MegaserverRU** [comment]
In the Crusader Trial, the same spel crashes the server on the 2nd boss Lord Jaraxus

**Comment by fullofstarschen** [comment]
@Nefertumm  
f0930f3 doesn't work either,
[d40e3340fb83+_worldserver.exe_[9-9_7-18-17].zip](https://github.com/azerothcore/azerothcore-wotlk/files/9531201/d40e3340fb83%2B_worldserver.exe_.9-9_7-18-17.zip)

**Comment by Nefertumm** [comment]
Try again, if this doesn't work I'm out of ideas

**Comment by MegaserverRU** [comment]
> Try again, if this doesn't work I'm out of ideas

Unfortunately this doesn't work either and causes a crash
https://github.com/azerothcore/azerothcore-wotlk/pull/12987/commits/1cf68e2c41fa9f462565a835791c92cfeae7c29c

**Comment by fullofstarschen** [comment]
1cf68e2c41fa9f462565a835791c92cfeae7c29c  crashed.
[f06f5a1334c4+_worldserver.exe_[19-9_7-3-26].zip](https://github.com/azerothcore/azerothcore-wotlk/files/9595456/f06f5a1334c4%2B_worldserver.exe_.19-9_7-3-26.zip)

**Comment by Nefertumm** [comment]
@fullofstarschen do you know how to reproduce it? I'll try to debug it.

**Comment by Nefertumm** [comment]
Nevermind, I know what's happening.

**Comment by fullofstarschen** [comment]
Use a hunter join in a raid and enter the Icecrown Citadel,choose a enemy cast some other spell and then cast Readiness,the world crashes.
@Nefertumm

**Comment by Nefertumm** [comment]
@fullofstarschen try now, it shouldn't crash anymore.

**Comment by Nefertumm** [comment]
To developers: I don't really like this hungry approach, using 2 sets is not the peak of efficiency. The right approach would be changing the methods that modifies the cooldown map, making them return an updated iterator position, but I don't know how to handle that with spell categories. 
TLDR: The crash happens because while we are looping caster->RemoveSpellCooldown(...) invalidates the iterator (it uses an erase method).
I'm really open to more efficient solutions, don't hesitate to ping me if you have one.

**Comment by fullofstarschen** [comment]
I tried 8aa7ad40d977e51239c9b7cbdd483555b9b3cbec and it seems to work,thank you @Nefertumm ,thank you @ZhengPeiRu21 @MegaserverRU @acidmanifesto

**Comment by MegaserverRU** [comment]
> @fullofstarschen попробуйте сейчас, он больше не должен зависать.

After a full check, it now works very well, no more failures were noticed
https://github.com/azerothcore/azerothcore-wotlk/pull/12987/commits/8aa7ad40d977e51239c9b7cbdd483555b9b3cbec
Thanks a lot [Nefertumm](https://github.com/Nefertumm)

**Comment by Nefertumm** [comment]
I'm going to rework this solution in a separated PR

**Comment by acidmanifesto** [comment]
> I'm going to rework this solution in a separated PR

so what is the disposition of this PR then? Close, Keep Open? Merge?

**Comment by Nefertumm** [comment]
When I open the new PR, I'll superseed this one closing it

**Comment by acidmanifesto** [comment]
i'll mark it as sensitive wip for now til you superscede it.

**Comment by Nefertumm** [comment]
Nevermind about superseding the PR. I won't be able to do it without restructuring everything. We better stick with this solution for now.
